### PR TITLE
Add MongodDB project

### PR DESCRIPTION
**Pull request by Rishi556**
`projection` allows for specific fields to be included or omitted in the response. Adding projection will allow for users to get just the data that they will be using, reducing bandwidth necessary. To use it, `project` can be added to the `contracts.find` and `contracts.findOne` calls within the query as an optional parameter(as well as internal calls). The default behavior is to return all fields. 

`rpcConfig.allowArbitraryProject` is added to the config, with a default value of false. If set to true, it allows project values to be whatever(can have calculations done on result before being sent back to client), but to not overwhelm public nodes, the default configuration just allows 1/0 values(ability to include/exclude fields).

Example call:

```sh
curl --location 'https://engine.rishipanthee.com' \
--header 'Content-Type: application/json' \
--data '{
  "jsonrpc": "2.0",
  "method": "contracts.find",
  "params": {"contract" : "nft", "table" : "STARinstances", "query" : {}, "project" : {"account": 1}},
  "id": 2
}'
```

### CODE REVIEW

File: libs/Database.js [hunk 1]
```diff
@@ -723,7 +736,7 @@ class Database {
             }
           }
 
-          result = await tableData.findOne(EJSON.deserialize(query), { session: this.session });
+          result = await tableData.findOne(EJSON.deserialize(query), { session: this.session, projection: project });
```
Comment (bt-cryptomancer) [inline libs/Database.js:739]: To be consistent, I feel like `findOne` should have the same sort of validation that is present in `find` above:

image

File: plugins/JsonRPCServer.js [hunk 2]
```diff
@@ -201,7 +201,15 @@ function contractsRPC() {
 
     findOne: async (args, callback) => {
       try {
-        const { contract, table, query } = args;
+        const { contract, table, query, project } = args;
```
Comment (bt-cryptomancer) [inline plugins/JsonRPCServer.js:204]: Should treat `project` parameter consistently as with `find` function below:

image

**Review by bt-cryptomancer — CHANGES_REQUESTED** [review]
See minor comments about data validation below.

Also, if you are adding support for projections to the RPC, you might as well add it to the version of these functions exposed to the smart contracts API as well, see `libs/SmartContracts.js`.  And then you'll also need a new unit test case (or add on to an existing test case for an arbitrary contract like, say, the tokens contract) to verify the modified smart contracts API works as expected if you include a project field.

File: plugins/JsonRPCServer.js (hunk 2 above)
Comment (Rishi556) [reply plugins/JsonRPCServer.js:209]: Completed in https://github.com/hive-engine/hivesmartcontracts/pull/19/commits/e58675510b80b60dc3986c338cffc14ed40be443

**Comment by Rishi556** [comment]
How would you say is the best way to expose/test it to the SC api? Add projection to a contract and use it there is what I'm thinking, but if you have a better idea, I'll look in that direction.

**Comment by bt-cryptomancer** [comment]
> How would you say is the best way to expose/test it to the SC api? Add projection to a contract and use it there is what I'm thinking, but if you have a better idea, I'll look in that direction.

This sounds like a reasonable approach.  What you could do is create a dummy test contract with one action that specifically utilizes this new feature and logs something on success.  Then have a unit test case that triggers the action and checks for the expected result in the contract logs.

**Comment by bt-cryptomancer** [comment]
@Rishi556 any updates on exposing this functionality to the smart contracts API as well as RPC?  I don't consider doing that very important, though, as we don't have any immediate planned smart contract uses for it.  So if you'd like to consider this task done for now, and hold off on smart contracts API support for another time, I'm fine with that.  Please confirm either way.

**Comment by Rishi556** [comment]
I tried writing a dummy contract with a test at one point, but it ended up being a bit messy and I haven't looked at it again since then. I think we can expose it to the smart contracts if its needed at another point, especially since nothing is using it as of now.

**Comment by bt-cryptomancer** [comment]
> I tried writing a dummy contract with a test at one point, but it ended up being a bit messy and I haven't looked at it again since then. I think we can expose it to the smart contracts if its needed at another point, especially since nothing is using it as of now.

Okay sounds good.  I'm happy to merge this as-is then, and consider it done for now.

**Review by bt-cryptomancer — APPROVED** [review]

**Pull request merged**